src/consensus/aft/raft.h

```diff
@@ -1183,10 +1164,56 @@ namespace aft
       {
         if (i <= state->last_idx)
         {
-          // If the current entry has already been deserialised, skip the
-          // payload for that entry
-          ledger->skip_entry(data, size);
-          continue;
+          // NB: This is only safe as long as AppendEntries only contain a
+          // single term. If they cover multiple terms, then we would need to at
+          // least partially deserialise each entry to establish what term it is
+          // in (or report the terms in the header)
+          static_assert(
+            max_terms_per_append_entries == 1,
+            "AppendEntries rollback logic assumes single term");
+          const auto incoming_term = r.term_of_idx;
+          const auto local_term = state->view_history.view_at(i);
+          if (incoming_term != local_term)
+          {
+            if (is_new_follower)
+            {
+              auto rollback_level = i - 1;
+              RAFT_DEBUG_FMT(
+                "New follower received AppendEntries with conflict. Incoming "
+                "entry {}.{} conflicts with local {}.{}. Rolling back to {}.",
+                incoming_term,
+                i,
+                local_term,
+                i,
+                rollback_level);
+              LOG_ROLLBACK_FMT(
+                "Dropping conflicting branch. Rolling back {} entries, "
+                "beginning with {}.{}.",
+                state->last_idx - rollback_level,
+                local_term,
+                i);
+              rollback(rollback_level);
+              is_new_follower = false;
+              // Then continue to process this AE as normal
+            }
+            else
+            {
+              LOG_ROLLBACK_FMT(
+                "Ignoring conflicting AppendEntries. Retaining {} entries, "
+                "beginning with {}.{}.",
+                state->last_idx - (i - 1),
+                local_term,
+                i);
+              return;
```

Comment:
For posterity: This branch is weird, and if we ever hit it it's a potential liveness hole. We have a node retaining a conflicting suffix, and refusing to roll it back. It will remain divergent (not contributing to commit) this term, and can only be brought in-sync in a future term.

A critical fix missed from earlier versions of this change is the `return` within this branch. If we hit this case, we log but then early-out, essentially ignoring the AE. We do not send the leader a NACK, since we would continue to ignore the resulting messages. If we're lucky, that may trigger `CheckQuorum` and a fresh election, avoiding a permanent liveness hole.

Reply:
> We have a node retaining a conflicting suffix, and refusing to roll it back. It will remain divergent (not contributing to commit) this term, and can only be brought in-sync in a future term.

That seems like a really good comment to have at :1201.